### ISSUE TITLE
Question about macro methods

### ISSUE THREAD

**Issue by Exilor** [manastech/crystal#1122]
In the [documentation](http://crystal-lang.org/docs/syntax_and_semantics/macros/macro_methods.html) there's a piece of code that is given as to demonstrate the power of macro methods. The thing is, the code appears to work just as well with a regular method definition: http://play.crystal-lang.org/#/r/ae4

This is more directed at @asterite:
In issue #1029 you told me that ```#to_h``` was broken and should be a macro def which I guess it's related to this. However, the method appears to work and other methods can be called on the resulting hash.

TL;DR I'd like to know what's the point of macro methods and what sets them apart of regular methods with ```{{}}``` code in them.

Bonus question:
What's the deal with certain macros not working unless enclosed by an ```if true```? I'm sure I've seen it before on this repo. Example [here](http://play.crystal-lang.org/#/r/ae8) and [here](http://play.crystal-lang.org/#/r/ae7).

**Comment by asterite** [comment]
Compare this:

http://play.crystal-lang.org/#/r/aeb

To this:

http://play.crystal-lang.org/#/r/aec

Once a method is instantiated, the compiler caches it and reuses it. In the first case the method was typed and the macro was evaluated, and the types at the moment were nil and nil. If you later change the types the method isn't recomputed.

Macro defs don't have these issues because their body is basically expanded after types stabilize. That's why you need to tell the compiler what's their type: because the compiler won't look into the body to infer that.

The above also behaves "wrong" in this case:

http://play.crystal-lang.org/#/r/aed

Here the compiler didn't know about the instance variable "@coco". But this is more controversial because maybe it could know it because it's used in a method.

**Comment by asterite** [comment]
Maybe all of this should be explained in the docs :-)

**Comment by kirbyfan64** [comment]
@asterite You *do* know that you just pinged a random GitHub user, right? :)

(I think you meant backticks instead of quotes, like `` `@coco` ``, which results in `@coco`?)

**Comment by asterite** [comment]
@kirbyfan64 Oops, you are right. I'll take that into account next time. Thanks!

**Comment by Exilor** [comment]
Oh I see... no wonder I couldn't figure it out on my own :laughing: thanks for the info!

**Comment by asterite** [comment]
@Exilor Totally missed that second quesiton. The docs say "The generated code must be valid Crystal code". The error message you get says "macro didn't expand to a valid program". And `0, 1, 2, 3, 4, 5,` isn't a valid Crystal program: http://play.crystal-lang.org/#/r/aep

And you can see that the error of that last program is the same as the error that the macro expansion is telling you.

So, the trick is to surround all of that (together with the brackets) with `{% if true %}`, to pass to "macro mode", and the total generated code becomes:

```crystal
[
0,
1,
2,
3,
4,
5,
] 
```

which is now a valid crystal program.

**Comment by Exilor** [comment]
Ahhh...! I thought the error message simply left out the array brackets. And I thought that the check for "valid Crystal code" was done after the macro had already "pasted" its contents.

That is all, thank you very much, closing this now :)
